System tests fail under emulator in clean environment

Similar to https://github.com/googleapis/python-bigtable/issues/431.

When the environment variable for the Firestore emulator is set, the client should disable any fallback lookups of credentials or project ID:  they aren't needed for running under the emulator.  At the moment, however:

```bash
$ env | grep CRED || echo "No credentials in environment"
No credentials in environment
$ env | grep PROJ || echo "No project in environment"
No project in environment
# Emulator running in separate shell -- no default creds available
$ which gcloud || echo "No gcloud"
No gcloud
$ export FIRESTORE_EMULATOR_HOST=localhost:8072
$ nox -s system -- -x
nox > Running session system-3.7
nox > Creating virtual environment (virtualenv) using python3.7 in .nox/system-3-7
nox > python -m pip install --pre grpcio
nox > python -m pip install mock pytest google-cloud-testutils pytest-asyncio -c /home/tseaver/projects/agendaless/Google/src/python-firestore/testing/constraints-3.7.txt
nox > python -m pip install -e . -c /home/tseaver/projects/agendaless/Google/src/python-firestore/testing/constraints-3.7.txt
nox > py.test --verbose --junitxml=system_3.7_sponge_log.xml tests/system -x
============================= test session starts ==============================
platform linux -- Python 3.7.12, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /home/tseaver/projects/agendaless/Google/src/python-firestore/.nox/system-3-7/bin/python
cachedir: .pytest_cache
rootdir: /home/tseaver/projects/agendaless/Google/src/python-firestore
plugins: asyncio-0.15.1
collected 94 items                                                             

tests/system/test_system.py::test_collections ERROR                      [  1%]

==================================== ERRORS ====================================
______________________ ERROR at setup of test_collections ______________________

    @pytest.fixture(scope="module")
    def client():
        credentials, project = _get_credentials_and_project()
>       yield firestore.Client(project=project, credentials=credentials)

tests/system/test_system.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
google/cloud/firestore_v1/client.py:99: in __init__
    client_options=client_options,
google/cloud/firestore_v1/base_client.py:129: in __init__
    _http=None,
.nox/system-3-7/lib/python3.7/site-packages/google/cloud/client.py:316: in __init__
    _ClientProjectMixin.__init__(self, project=project, credentials=credentials)
.nox/system-3-7/lib/python3.7/site-packages/google/cloud/client.py:264: in __init__
    project = self._determine_default(project)
.nox/system-3-7/lib/python3.7/site-packages/google/cloud/client.py:283: in _determine_default
    return _determine_default_project(project)
.nox/system-3-7/lib/python3.7/site-packages/google/cloud/_helpers.py:152: in _determine_default_project
    _, project = google.auth.default()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

scopes = None, request = None, quota_project_id = None, default_scopes = None

    def default(scopes=None, request=None, quota_project_id=None, default_scopes=None):
        """Gets the default credentials for the current environment.
        ...
>       raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
E       google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started

.nox/system-3-7/lib/python3.7/site-packages/google/auth/_default.py:488: DefaultCredentialsError
...
=============================== 1 error in 6.43s ===============================
nox > Command py.test --verbose --junitxml=system_3.7_sponge_log.xml tests/system -x failed with exit code 1
nox > Session system-3.7 failed.
```

--- a/.github/workflows/system_emulated.yml
+++ b/.github/workflows/system_emulated.yml
@@ -1,0 +1,30 @@
+name: "Run systests on emulator"
+on:
+ #pull_request
+ #  branches:
+ #    - main
+  push:  # temporary
+
+jobs:
+
+  run-systests:
+    runs-on: ubuntu-20.04
+
+    steps:
+
+    - name: Checkout
+      uses: actions/checkout@v2
+
+    - name: Setup Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.8'
+
+    - name: Setup GCloud SDK
+      uses: google-github-actions/setup-gcloud@v0.2.1
+
+    - name: Install / run Nox
+      run: |
+        python -m pip install --upgrade setuptools pip
+        python -m pip install nox
+        nox -s system_emulated

--- a/.github/workflows/system_emulated.yml
+++ b/.github/workflows/system_emulated.yml
@@ -18,7 +18,7 @@ jobs:
     - name: Setup Python
       uses: actions/setup-python@v2
       with:
-        python-version: '3.8'
+        python-version: '3.7'
 
     - name: Setup GCloud SDK
       uses: google-github-actions/setup-gcloud@v0.2.1

--- a/.github/workflows/system_emulated.yml
+++ b/.github/workflows/system_emulated.yml
@@ -1,9 +1,8 @@
 name: "Run systests on emulator"
 on:
- #pull_request
- #  branches:
- #    - main
-  push:  # temporary
+  pull_request:
+    branches:
+      - main
 
 jobs:
 

--- a/google/cloud/firestore_v1/base_client.py
+++ b/google/cloud/firestore_v1/base_client.py
@@ -62,6 +62,7 @@ from google.cloud.firestore_v1.base_query import BaseQuery
 
 DEFAULT_DATABASE = "(default)"
 """str: The default database used in a :class:`~google.cloud.firestore_v1.client.Client`."""
+_DEFAULT_EMULATOR_PROJECT = "google-cloud-firestore-emulator"
 _BAD_OPTION_ERR = (
     "Exactly one of ``last_update_time`` or ``exists`` " "must be provided."
 )
@@ -129,7 +130,7 @@ class BaseClient(ClientWithProject):
             if credentials is None:
                 credentials = AnonymousCredentials()
             if project is None:
-                project = "nonesuch-project-123"
+                project = _DEFAULT_EMULATOR_PROJECT
 
         super(BaseClient, self).__init__(
             project=project,

--- a/google/cloud/firestore_v1/base_client.py
+++ b/google/cloud/firestore_v1/base_client.py
@@ -27,6 +27,7 @@ In the hierarchy of API concepts
 import os
 import grpc  # type: ignore
 
+from google.auth.credentials import AnonymousCredentials
 import google.api_core.client_options  # type: ignore
 import google.api_core.path_template  # type: ignore
 from google.api_core import retry as retries  # type: ignore
@@ -122,6 +123,14 @@ class BaseClient(ClientWithProject):
         # NOTE: This API has no use for the _http argument, but sending it
         #       will have no impact since the _http() @property only lazily
         #       creates a working HTTP object.
+        self._emulator_host = os.getenv(_FIRESTORE_EMULATOR_HOST)
+
+        if self._emulator_host is not None:
+            if credentials is None:
+                credentials = AnonymousCredentials()
+            if project is None:
+                project = "nonesuch-project-123"
+
         super(BaseClient, self).__init__(
             project=project,
             credentials=credentials,
@@ -137,7 +146,6 @@ class BaseClient(ClientWithProject):
         self._client_options = client_options
 
         self._database = database
-        self._emulator_host = os.getenv(_FIRESTORE_EMULATOR_HOST)
 
     def _firestore_api_helper(self, transport, client_class, client_module) -> Any:
         """Lazy-loading getter GAPIC Firestore API.

--- a/noxfile.py
+++ b/noxfile.py
@@ -37,6 +37,7 @@ CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()
 # 'docfx' is excluded since it only needs to run in 'docs-presubmit'
 nox.options.sessions = [
     "unit",
+    "system_emulated",
     "system",
     "cover",
     "lint",
@@ -126,6 +127,33 @@ def default(session):
 def unit(session):
     """Run the unit test suite."""
     default(session)
+
+
+@nox.session(python="3.8")
+def system_emulated(session):
+    import subprocess
+    import signal
+
+    try:
+        subprocess.call(["gcloud", "--version"])
+    except OSError:
+        session.skip("gcloud not found but required for emulator support")
+
+    # Currently, CI/CD doesn't have beta component of gcloud.
+    subprocess.call(["gcloud", "components", "install", "beta", "firestore"])
+
+    hostport = "localhost:8789"
+    session.env["FIRESTORE_EMULATOR_HOST"] = hostport
+
+    p = subprocess.Popen(
+        ["gcloud", "beta", "emulators", "firestore", "start", "--host-port", hostport]
+    )
+
+    try:
+        system(session)
+    finally:
+        # Stop Emulator
+        os.killpg(os.getpgid(p.pid), signal.SIGKILL)
 
 
 @nox.session(python=SYSTEM_TEST_PYTHON_VERSIONS)

--- a/noxfile.py
+++ b/noxfile.py
@@ -140,7 +140,9 @@ def system_emulated(session):
         session.skip("gcloud not found but required for emulator support")
 
     # Currently, CI/CD doesn't have beta component of gcloud.
-    subprocess.call(["gcloud", "components", "install", "beta", "firestore"])
+    subprocess.call(
+        ["gcloud", "components", "install", "beta", "cloud-firestore-emulator",]
+    )
 
     hostport = "localhost:8789"
     session.env["FIRESTORE_EMULATOR_HOST"] = hostport

--- a/noxfile.py
+++ b/noxfile.py
@@ -164,8 +164,7 @@ def system_emulated(session):
         system(session)
     finally:
         # Stop Emulator
-        # os.killpg(os.getpgid(p.pid), signal.SIGKILL)
-        pass
+        os.killpg(os.getpgid(p.pid), signal.SIGKILL)
 
 
 @nox.session(python=SYSTEM_TEST_PYTHON_VERSIONS)

--- a/noxfile.py
+++ b/noxfile.py
@@ -162,7 +162,8 @@ def system_emulated(session):
         system(session)
     finally:
         # Stop Emulator
-        os.killpg(os.getpgid(p.pid), signal.SIGKILL)
+        # os.killpg(os.getpgid(p.pid), signal.SIGKILL)
+        pass
 
 
 @nox.session(python=SYSTEM_TEST_PYTHON_VERSIONS)

--- a/noxfile.py
+++ b/noxfile.py
@@ -129,7 +129,7 @@ def unit(session):
     default(session)
 
 
-@nox.session(python="3.8")
+@nox.session(python=SYSTEM_TEST_PYTHON_VERSIONS)
 def system_emulated(session):
     import subprocess
     import signal

--- a/noxfile.py
+++ b/noxfile.py
@@ -146,7 +146,16 @@ def system_emulated(session):
     session.env["FIRESTORE_EMULATOR_HOST"] = hostport
 
     p = subprocess.Popen(
-        ["gcloud", "beta", "emulators", "firestore", "start", "--host-port", hostport]
+        [
+            "gcloud",
+            "--quiet",
+            "beta",
+            "emulators",
+            "firestore",
+            "start",
+            "--host-port",
+            hostport,
+        ]
     )
 
     try:

--- a/owlbot.py
+++ b/owlbot.py
@@ -155,7 +155,7 @@ def place_before(path, text, *before_text, escape=None):
     s.replace([path], text, replacement)
 
 system_emulated_session = """
-@nox.session(python="3.8")
+@nox.session(python=SYSTEM_TEST_PYTHON_VERSIONS)
 def system_emulated(session):
     import subprocess
     import signal

--- a/owlbot.py
+++ b/owlbot.py
@@ -190,8 +190,7 @@ def system_emulated(session):
         system(session)
     finally:
         # Stop Emulator
-        # os.killpg(os.getpgid(p.pid), signal.SIGKILL)
-        pass
+        os.killpg(os.getpgid(p.pid), signal.SIGKILL)
 
 """
 

--- a/owlbot.py
+++ b/owlbot.py
@@ -188,7 +188,8 @@ def system_emulated(session):
         system(session)
     finally:
         # Stop Emulator
-        os.killpg(os.getpgid(p.pid), signal.SIGKILL)
+        # os.killpg(os.getpgid(p.pid), signal.SIGKILL)
+        pass
 
 """
 

--- a/owlbot.py
+++ b/owlbot.py
@@ -166,7 +166,9 @@ def system_emulated(session):
         session.skip("gcloud not found but required for emulator support")
 
     # Currently, CI/CD doesn't have beta component of gcloud.
-    subprocess.call(["gcloud", "components", "install", "beta", "firestore"])
+    subprocess.call(
+        ["gcloud", "components", "install", "beta", "cloud-firestore-emulator",]
+    )
 
     hostport = "localhost:8789"
     session.env["FIRESTORE_EMULATOR_HOST"] = hostport

--- a/owlbot.py
+++ b/owlbot.py
@@ -143,6 +143,65 @@ python.py_samples(skip_readmes=True)
 
 s.move(templated_files)
 
+# ----------------------------------------------------------------------------
+# Customize noxfile.py
+# ----------------------------------------------------------------------------
+
+def place_before(path, text, *before_text, escape=None):
+    replacement = "\n".join(before_text) + "\n" + text
+    if escape:
+        for c in escape:
+            text = text.replace(c, '\\' + c)
+    s.replace([path], text, replacement)
+
+system_emulated_session = """
+@nox.session(python="3.8")
+def system_emulated(session):
+    import subprocess
+    import signal
+
+    try:
+        subprocess.call(["gcloud", "--version"])
+    except OSError:
+        session.skip("gcloud not found but required for emulator support")
+
+    # Currently, CI/CD doesn't have beta component of gcloud.
+    subprocess.call(["gcloud", "components", "install", "beta", "firestore"])
+
+    hostport = "localhost:8789"
+    session.env["FIRESTORE_EMULATOR_HOST"] = hostport
+
+    p = subprocess.Popen(
+        ["gcloud", "beta", "emulators", "firestore", "start", "--host-port", hostport]
+    )
+
+    try:
+        system(session)
+    finally:
+        # Stop Emulator
+        os.killpg(os.getpgid(p.pid), signal.SIGKILL)
+
+"""
+
+place_before(
+    "noxfile.py",
+    "@nox.session(python=SYSTEM_TEST_PYTHON_VERSIONS)\n"
+    "def system(session):",
+    system_emulated_session,
+    escape="()"
+)
+
+# add system_emulated nox session
+s.replace("noxfile.py",
+    """nox.options.sessions = \[
+    "unit",
+    "system",""",
+    """nox.options.sessions = [
+    "unit",
+    "system_emulated",
+    "system",""",
+)
+
 s.replace(
     "noxfile.py",
     """\"--quiet\",

--- a/owlbot.py
+++ b/owlbot.py
@@ -172,7 +172,16 @@ def system_emulated(session):
     session.env["FIRESTORE_EMULATOR_HOST"] = hostport
 
     p = subprocess.Popen(
-        ["gcloud", "beta", "emulators", "firestore", "start", "--host-port", hostport]
+        [
+            "gcloud",
+            "--quiet",
+            "beta",
+            "emulators",
+            "firestore",
+            "start",
+            "--host-port",
+            hostport,
+        ]
     )
 
     try:

--- a/tests/unit/v1/test_async_client.py
+++ b/tests/unit/v1/test_async_client.py
@@ -51,18 +51,6 @@ class TestAsyncClient(aiounittest.AsyncTestCase):
         self.assertEqual(client._credentials, credentials)
         self.assertEqual(client._database, DEFAULT_DATABASE)
         self.assertIs(client._client_info, _CLIENT_INFO)
-        self.assertIsNone(client._emulator_host)
-
-    def test_constructor_with_emulator_host(self):
-        from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
-
-        credentials = _make_credentials()
-        emulator_host = "localhost:8081"
-        with mock.patch("os.getenv") as getenv:
-            getenv.return_value = emulator_host
-            client = self._make_one(project=self.PROJECT, credentials=credentials)
-            self.assertEqual(client._emulator_host, emulator_host)
-            getenv.assert_called_once_with(_FIRESTORE_EMULATOR_HOST)
 
     def test_constructor_explicit(self):
         from google.api_core.client_options import ClientOptions

--- a/tests/unit/v1/test_base_client.py
+++ b/tests/unit/v1/test_base_client.py
@@ -41,7 +41,6 @@ class TestBaseClient(unittest.TestCase):
         from google.auth.credentials import AnonymousCredentials
         from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
 
-        credentials = _make_credentials()
         emulator_host = "localhost:8081"
 
         with mock.patch("os.environ", {_FIRESTORE_EMULATOR_HOST: emulator_host}):
@@ -55,7 +54,6 @@ class TestBaseClient(unittest.TestCase):
         from google.auth.credentials import AnonymousCredentials
         from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
 
-        credentials = _make_credentials()
         emulator_host = "localhost:8081"
 
         with mock.patch("os.environ", {_FIRESTORE_EMULATOR_HOST: emulator_host}):

--- a/tests/unit/v1/test_base_client.py
+++ b/tests/unit/v1/test_base_client.py
@@ -37,6 +37,46 @@ class TestBaseClient(unittest.TestCase):
         credentials = _make_credentials()
         return self._make_one(project=self.PROJECT, credentials=credentials)
 
+    def test_constructor_with_emulator_host_defaults(self):
+        from google.auth.credentials import AnonymousCredentials
+        from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
+
+        credentials = _make_credentials()
+        emulator_host = "localhost:8081"
+
+        with mock.patch("os.environ", {_FIRESTORE_EMULATOR_HOST: emulator_host}):
+            client = self._make_one()
+
+        self.assertEqual(client._emulator_host, emulator_host)
+        self.assertIsInstance(client._credentials, AnonymousCredentials)
+        self.assertEqual(client.project, "nonesuch-project-123")
+
+    def test_constructor_with_emulator_host_w_project(self):
+        from google.auth.credentials import AnonymousCredentials
+        from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
+
+        credentials = _make_credentials()
+        emulator_host = "localhost:8081"
+
+        with mock.patch("os.environ", {_FIRESTORE_EMULATOR_HOST: emulator_host}):
+            client = self._make_one(project=self.PROJECT)
+
+        self.assertEqual(client._emulator_host, emulator_host)
+        self.assertIsInstance(client._credentials, AnonymousCredentials)
+
+    def test_constructor_with_emulator_host_w_creds(self):
+        from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
+
+        credentials = _make_credentials()
+        emulator_host = "localhost:8081"
+
+        with mock.patch("os.environ", {_FIRESTORE_EMULATOR_HOST: emulator_host}):
+            client = self._make_one(credentials=credentials)
+
+        self.assertEqual(client._emulator_host, emulator_host)
+        self.assertIs(client._credentials, credentials)
+        self.assertEqual(client.project, "nonesuch-project-123")
+
     @mock.patch(
         "google.cloud.firestore_v1.services.firestore.client.FirestoreClient",
         autospec=True,

--- a/tests/unit/v1/test_base_client.py
+++ b/tests/unit/v1/test_base_client.py
@@ -39,6 +39,7 @@ class TestBaseClient(unittest.TestCase):
 
     def test_constructor_with_emulator_host_defaults(self):
         from google.auth.credentials import AnonymousCredentials
+        from google.cloud.firestore_v1.base_client import _DEFAULT_EMULATOR_PROJECT
         from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
 
         emulator_host = "localhost:8081"
@@ -48,7 +49,7 @@ class TestBaseClient(unittest.TestCase):
 
         self.assertEqual(client._emulator_host, emulator_host)
         self.assertIsInstance(client._credentials, AnonymousCredentials)
-        self.assertEqual(client.project, "nonesuch-project-123")
+        self.assertEqual(client.project, _DEFAULT_EMULATOR_PROJECT)
 
     def test_constructor_with_emulator_host_w_project(self):
         from google.auth.credentials import AnonymousCredentials
@@ -63,6 +64,7 @@ class TestBaseClient(unittest.TestCase):
         self.assertIsInstance(client._credentials, AnonymousCredentials)
 
     def test_constructor_with_emulator_host_w_creds(self):
+        from google.cloud.firestore_v1.base_client import _DEFAULT_EMULATOR_PROJECT
         from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
 
         credentials = _make_credentials()
@@ -73,7 +75,7 @@ class TestBaseClient(unittest.TestCase):
 
         self.assertEqual(client._emulator_host, emulator_host)
         self.assertIs(client._credentials, credentials)
-        self.assertEqual(client.project, "nonesuch-project-123")
+        self.assertEqual(client.project, _DEFAULT_EMULATOR_PROJECT)
 
     @mock.patch(
         "google.cloud.firestore_v1.services.firestore.client.FirestoreClient",

--- a/tests/unit/v1/test_client.py
+++ b/tests/unit/v1/test_client.py
@@ -49,18 +49,6 @@ class TestClient(unittest.TestCase):
         self.assertEqual(client._credentials, credentials)
         self.assertEqual(client._database, DEFAULT_DATABASE)
         self.assertIs(client._client_info, _CLIENT_INFO)
-        self.assertIsNone(client._emulator_host)
-
-    def test_constructor_with_emulator_host(self):
-        from google.cloud.firestore_v1.base_client import _FIRESTORE_EMULATOR_HOST
-
-        credentials = _make_credentials()
-        emulator_host = "localhost:8081"
-        with mock.patch("os.getenv") as getenv:
-            getenv.return_value = emulator_host
-            client = self._make_one(project=self.PROJECT, credentials=credentials)
-            self.assertEqual(client._emulator_host, emulator_host)
-            getenv.assert_called_once_with(_FIRESTORE_EMULATOR_HOST)
 
     def test_constructor_explicit(self):
         from google.api_core.client_options import ClientOptions

